File: legacy-tools/fixtures/real-project-surface-verdict.mjs
import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

export const realProjectSurfaceNames = [
  "waiver-audit",
  "typecheck-dependencies",
  "core-tools",
  "lsp",
  "lint-divergence",
  "syntax-highlighter",
  "glyph",
  "typecheck-divergence",
];

const validOutcomes = new Set(["success", "failure", "cancelled", "skipped"]);

function recordOnlyVerdict(outcome, mode) {
  return mode === "record-only" && outcome === "failure" ? "success" : outcome;
}

export function createRealProjectSurfaceResultsFromWorkflow(environment = process.env) {
  return [
    { name: "waiver-audit", outcome: environment.VIZE_WAIVER_AUDIT_OUTCOME },
    {
      name: "typecheck-dependencies",
      outcome: recordOnlyVerdict(
        environment.VIZE_TYPECHECK_DEPENDENCIES_OUTCOME,
        environment.TYPECHECK_DEPENDENCIES_MODE,
      ),
    },
    {
      name: "core-tools",
      outcome: recordOnlyVerdict(environment.VIZE_CORE_TOOLS_OUTCOME, environment.CORE_TOOLS_MODE),
    },
    {
      name: "lsp",
      outcome: recordOnlyVerdict(environment.VIZE_LSP_OUTCOME, environment.LSP_MODE),
    },
    {
      name: "lint-divergence",
      outcome: recordOnlyVerdict(
        environment.VIZE_LINT_DIVERGENCE_OUTCOME,
        environment.LINT_DIVERGENCE_MODE,
      ),
    },
    { name: "syntax-highlighter", outcome: environment.VIZE_SYNTAX_HIGHLIGHTER_OUTCOME },
    { name: "glyph", outcome: environment.VIZE_GLYPH_OUTCOME },
    {
      name: "typecheck-divergence",
      outcome: recordOnlyVerdict(
        environment.VIZE_TYPECHECK_DIVERGENCE_OUTCOME,
        environment.TYPECHECK_DIVERGENCE_MODE,
      ),
    },
  ];
}

export function createRealProjectSurfaceVerdict(results, environment = process.env) {
  const expected = new Set(realProjectSurfaceNames);
  const seen = new Set();
  for (const result of results) {
    if (result == null || typeof result !== "object" || Array.isArray(result)) {
      throw new Error("surface results must be objects");
    }
    if (!expected.has(result.name)) {
      throw new Error(`unknown real-project surface: ${String(result.name)}`);
    }
    if (seen.has(result.name)) {
      throw new Error(`duplicate real-project surface: ${result.name}`);
    }
    if (!validOutcomes.has(result.outcome)) {
      throw new Error(`invalid outcome for ${result.name}: ${String(result.outcome)}`);
    }
    seen.add(result.name);
  }
  const missing = realProjectSurfaceNames.filter((name) => !seen.has(name));
  if (missing.length > 0) {
    throw new Error(`missing real-project surface verdict(s): ${missing.join(", ")}`);
  }
  const failed = results.filter((result) => result.outcome !== "success");
  return {
    schema: "vize.realProjectSurfaceVerdict",
    version: 1,
    sourceCommit: environment.GITHUB_SHA ?? null,
    shardIndex: environment.FIXTURE_SHARD_INDEX ?? null,
    status: failed.length === 0 ? "success" : "failure",
    surfaces: realProjectSurfaceNames.map((name) => results.find((result) => result.name === name)),
    failedSurfaceNames: failed
      .map((result) => result.name)
      .sort((left, right) => left.localeCompare(right)),
  };
}

function parseArguments(argv) {
  let output = null;
  let fromWorkflowEnv = false;
  const results = [];
  for (let index = 0; index < argv.length; index += 1) {
    const argument = argv[index];
    const value = argv[index + 1];
    if (argument === "--from-workflow-env") {
      fromWorkflowEnv = true;
      continue;
    }
    if (argument === "--output" && output == null && value && !value.startsWith("--")) {
      output = resolve(value);
      index += 1;
      continue;
    }
    if (argument === "--surface" && value && !value.startsWith("--")) {
      const separator = value.indexOf("=");
      if (separator < 1) throw new Error(`invalid --surface value: ${value}`);
      results.push({ name: value.slice(0, separator), outcome: value.slice(separator + 1) });
      index += 1;
      continue;
    }
    throw new Error(`unknown or incomplete argument: ${argument}`);
  }
  if (output == null) throw new Error("--output is required");
  if (fromWorkflowEnv && results.length > 0) {
    throw new Error("--from-workflow-env cannot be combined with --surface");
  }
  return { output, fromWorkflowEnv, results };
}

function main() {
  const { output, fromWorkflowEnv, results } = parseArguments(process.argv.slice(2));
  const surfaces = fromWorkflowEnv ? createRealProjectSurfaceResultsFromWorkflow() : results;
  const artifact = createRealProjectSurfaceVerdict(surfaces);
  mkdirSync(dirname(output), { recursive: true });
  writeFileSync(output, `${JSON.stringify(artifact, null, 2)}\n`);
  if (artifact.status !== "success") {
    throw new Error(`real-project surfaces failed: ${artifact.failedSurfaceNames.join(", ")}`);
  }
  process.stdout.write(`all ${artifact.surfaces.length} real-project surfaces succeeded\n`);
}

const isMain = process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (isMain) {
  try {
    main();
  } catch (error) {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
    process.exitCode = 1;
  }
}

File: tools/commands/fixtures/real-project-surface-verdict.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../rust/common.rs"]
mod common;

use serde::Serialize;
use std::{collections::BTreeSet, env, path::PathBuf, process::ExitCode};

const SURFACE_NAMES: &[&str] = &[
    "waiver-audit",
    "typecheck-dependencies",
    "core-tools",
    "lsp",
    "lint-divergence",
    "syntax-highlighter",
    "glyph",
    "typecheck-divergence",
];

#[derive(Clone, Debug, PartialEq, Serialize)]
struct SurfaceResult {
    name: String,
    outcome: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct Verdict {
    schema: &'static str,
    version: u8,
    source_commit: Option<String>,
    shard_index: Option<String>,
    status: &'static str,
    surfaces: Vec<SurfaceResult>,
    failed_surface_names: Vec<String>,
}

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let args = parse_args(env::args().skip(1))?;
    let surfaces = if args.from_workflow_env {
        results_from_workflow_env()
    } else {
        args.results
    };
    let verdict = create_verdict(surfaces)?;
    common::write_json_pretty(&args.output, &verdict)?;
    if verdict.status != "success" {
        return Err(format!(
            "real-project surfaces failed: {}",
            verdict.failed_surface_names.join(", ")
        ));
    }
    println!(
        "all {} real-project surfaces succeeded",
        verdict.surfaces.len()
    );
    Ok(())
}

struct Args {
    output: PathBuf,
    from_workflow_env: bool,
    results: Vec<SurfaceResult>,
}

fn parse_args(args: impl Iterator<Item = String>) -> Result<Args, String> {
    let mut output = None;
    let mut from_workflow_env = false;
    let mut results = Vec::new();
    let mut args = args.peekable();
    while let Some(argument) = args.next() {
        match argument.as_str() {
            "--from-workflow-env" => from_workflow_env = true,
            "--output" => {
                let value = args
                    .next()
                    .filter(|value| !value.starts_with("--"))
                    .ok_or_else(|| "--output requires a value".to_string())?;
                output = Some(PathBuf::from(value));
            }
            "--surface" => {
                let value = args
                    .next()
                    .filter(|value| !value.starts_with("--"))
                    .ok_or_else(|| "--surface requires a value".to_string())?;
                let Some((name, outcome)) = value.split_once('=') else {
                    return Err(format!("invalid --surface value: {value}"));
                };
                if name.is_empty() {
                    return Err(format!("invalid --surface value: {value}"));
                }
                results.push(SurfaceResult {
                    name: name.to_string(),
                    outcome: outcome.to_string(),
                });
            }
            _ => return Err(format!("unknown or incomplete argument: {argument}")),
        }
    }
    if output.is_none() {
        return Err("--output is required".to_string());
    }
    if from_workflow_env && !results.is_empty() {
        return Err("--from-workflow-env cannot be combined with --surface".to_string());
    }
    Ok(Args {
        output: output.unwrap(),
        from_workflow_env,
        results,
    })
}

fn results_from_workflow_env() -> Vec<SurfaceResult> {
    vec![
        result("waiver-audit", env_value("VIZE_WAIVER_AUDIT_OUTCOME")),
        result(
            "typecheck-dependencies",
            record_only_verdict(
                env_value("VIZE_TYPECHECK_DEPENDENCIES_OUTCOME"),
                env_value("TYPECHECK_DEPENDENCIES_MODE"),
            ),
        ),
        result(
            "core-tools",
            record_only_verdict(
                env_value("VIZE_CORE_TOOLS_OUTCOME"),
                env_value("CORE_TOOLS_MODE"),
            ),
        ),
        result(
            "lsp",
            record_only_verdict(env_value("VIZE_LSP_OUTCOME"), env_value("LSP_MODE")),
        ),
        result(
            "lint-divergence",
            record_only_verdict(
                env_value("VIZE_LINT_DIVERGENCE_OUTCOME"),
                env_value("LINT_DIVERGENCE_MODE"),
            ),
        ),
        result(
            "syntax-highlighter",
            env_value("VIZE_SYNTAX_HIGHLIGHTER_OUTCOME"),
        ),
        result("glyph", env_value("VIZE_GLYPH_OUTCOME")),
        result(
            "typecheck-divergence",
            record_only_verdict(
                env_value("VIZE_TYPECHECK_DIVERGENCE_OUTCOME"),
                env_value("TYPECHECK_DIVERGENCE_MODE"),
            ),
        ),
    ]
}

fn result(name: &str, outcome: String) -> SurfaceResult {
    SurfaceResult {
        name: name.to_string(),
        outcome,
    }
}

fn env_value(name: &str) -> String {
    env::var(name).unwrap_or_default()
}

fn record_only_verdict(outcome: String, mode: String) -> String {
    if mode == "record-only" && outcome == "failure" {
        "success".to_string()
    } else {
        outcome
    }
}

fn create_verdict(results: Vec<SurfaceResult>) -> Result<Verdict, String> {
    let expected = SURFACE_NAMES.iter().copied().collect::<BTreeSet<_>>();
    let mut seen = BTreeSet::new();
    for result in &results {
        if !expected.contains(result.name.as_str()) {
            return Err(format!("unknown real-project surface: {}", result.name));
        }
        if !seen.insert(result.name.as_str()) {
            return Err(format!("duplicate real-project surface: {}", result.name));
        }
        if !matches!(
            result.outcome.as_str(),
            "success" | "failure" | "cancelled" | "skipped"
        ) {
            return Err(format!(
                "invalid outcome for {}: {}",
                result.name, result.outcome
            ));
        }
    }
    let missing = SURFACE_NAMES
        .iter()
        .copied()
        .filter(|name| !seen.contains(name))
        .collect::<Vec<_>>();
    if !missing.is_empty() {
        return Err(format!(
            "missing real-project surface verdict(s): {}",
            missing.join(", ")
        ));
    }
    let failed = results
        .iter()
        .filter(|result| result.outcome != "success")
        .map(|result| result.name.clone())
        .collect::<Vec<_>>();
    let mut failed_sorted = failed;
    failed_sorted.sort();
    Ok(Verdict {
        schema: "vize.realProjectSurfaceVerdict",
        version: 1,
        source_commit: env::var("GITHUB_SHA").ok(),
        shard_index: env::var("FIXTURE_SHARD_INDEX").ok(),
        status: if failed_sorted.is_empty() {
            "success"
        } else {
            "failure"
        },
        surfaces: SURFACE_NAMES
            .iter()
            .filter_map(|name| results.iter().find(|result| result.name == *name).cloned())
            .collect(),
        failed_surface_names: failed_sorted,
    })
}
